## Shipment and delivery delay analysis


Let's explore the **orders** and **order_items** datasets to verify the efficiency of the shipment and delivery of the platform by answering the following questions:

1. How many orders were shipped with delay?
2. What are the regions with most delays?
3. Are there any seller with high delay rate?
4. How many orders were delivered with delay?
5. How shipping delay impacts delivery performance?
6. Are there any relation between delivery delay and the distance between sellers and customers?
7. How does freight value behaves among states?
8. Conclusions

# 1. How many orders were shipped with delay?

Let's begin with the order shipment efficiency. To do so, we will consider all orders that have a date of delivering to a carrier (represented by the column "order_delivered_carrier_date").

In [1]:
import os
import pandas as pd
import mysql.connector as mysql
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.lines as mlines
from credentials import login_config, path_dir
import datetime as dt

from bokeh.plotting import figure, show
from bokeh.io import output_notebook, curdoc, push_notebook
from bokeh.models import ColumnDataSource, HoverTool, LinearAxis, Range1d, FactorRange
from bokeh.layouts import column, row
from bokeh.models.widgets import Select

output_notebook()

Loading BokehJS ...

In [2]:
# Adjusting the working directory to better manage the file calls
os.chdir(path_dir)

In [3]:
# Report information with order dispatching efficiency per month

# Creating a connection to the MySQL Database
connection = mysql.connect(**login_config)

# parsing the period of evaluation
period = ('2016-01-01 00:00:00', '2018-12-31 23:59:59')

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/delayed_shipments_by_year_month.sql") as query:
    order_shipment_delay_raw = pd.read_sql_query(query.read(), params=period, con=connection)

order_shipment_delay_raw

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


year_months  Shipped_on_schedule  Shipped_with_delay  total_shipments  \
0        201610                167.0               129.0              296   
1        201611                  1.0                35.0               36   
2        201612                  1.0                 1.0                2   
3        201701                598.0               128.0              726   
4        201702               1424.0               295.0             1719   
5        201703               2813.0               292.0             3105   
6        201704               2241.0               147.0             2388   
7        201705               3848.0               331.0             4179   
8        201706               3339.0               304.0             3643   
9        201707               3871.0               329.0             4200   
10       201708               4582.0               376.0             4958   
11       201709               4350.0               341.0             4691   
12       201710               4821.0               392.0             5213   
13       201711               7098.0               611.0             7709   
14       201712               5787.0              1098.0             6885   
15       201801               7099.0               861.0             7960   
16       201802               6594.0               955.0             7549   
17       201803               7480.0               786.0             8266   
18       201804               7446.0               597.0             8043   
19       201805               7570.0               518.0             8088   
20       201806               6551.0               457.0             7008   
21       201807               6212.0               595.0             6807   
22       201808               7140.0               843.0             7983   
23       201809                  1.0                 2.0                3   

    Delay_Percentage  Average_Delay  Average_Shipping_time  \
0              43.58          7.909                 11.421   
1              97.22         28.298                 37.671   
2              50.00         58.592                 33.051   
3              17.63          1.994                  2.676   
4              17.16          3.629                  3.206   
5               9.40          5.565                  3.404   
6               6.16          4.181                  3.400   
7               7.92          2.938                  3.346   
8               8.34          2.792                  3.170   
9               7.83          1.782                  3.118   
10              7.58          2.358                  3.127   
11              7.27          2.508                  3.191   
12              7.52          3.064                  3.340   
13              7.93          2.525                  3.576   
14             15.95          3.281                  4.024   
15             10.82          4.188                  3.750   
16             12.65          3.065                  3.886   
17              9.51          3.487                  3.359   
18              7.42          2.567                  3.049   
19              6.40          2.142                  2.701   
20              6.52          2.101                  2.598   
21              8.74          1.959                  2.812   
22             10.56          2.491                  2.909   
23             66.67         23.864                 20.669   

    Average_Shipping_estimate  
0                       8.214  
1                       9.489  
2                       4.109  
3                       4.559  
4                       5.196  
5                       7.431  
6                       7.992  
7                       7.768  
8                       7.027  
9                       6.858  
10                      6.831  
11                      7.012  
12                      7.139  
13                      7.015  
14                      7.003  
15             

It is visible that the orders from 2016 and september-2018 had the biggest delays. Apart from those months, the rest of the period had a percentage of delayed orders inferior to 20%.

Another interesting point to take is the average delay in days. This data (calculated by the difference between the shipping date and the shipping limit) show that the months with most percentages of orders delayed also had the most delay in days.
Since this data are in the edges of the dataset and have low representation (0,35% of total orders) i will opt to truncate this data for the rest of the analysis.

In [4]:
# Recalling the query using the truncated period

# parsing the period of evaluation
period = ('2017-01-01 00:00:00', '2018-08-31 23:59:59')

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/delayed_shipments_by_year_month.sql") as query:
    order_shipment_delay_per_year_month = pd.read_sql_query(query.read(), params=period, con=connection)

order_shipment_delay_per_year_month

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


year_months  Shipped_on_schedule  Shipped_with_delay  total_shipments  \
0        201701                598.0               128.0              726   
1        201702               1424.0               295.0             1719   
2        201703               2813.0               292.0             3105   
3        201704               2241.0               147.0             2388   
4        201705               3848.0               331.0             4179   
5        201706               3339.0               304.0             3643   
6        201707               3871.0               329.0             4200   
7        201708               4582.0               376.0             4958   
8        201709               4350.0               341.0             4691   
9        201710               4821.0               392.0             5213   
10       201711               7098.0               611.0             7709   
11       201712               5787.0              1098.0             6885   
12       201801               7099.0               861.0             7960   
13       201802               6594.0               955.0             7549   
14       201803               7480.0               786.0             8266   
15       201804               7446.0               597.0             8043   
16       201805               7570.0               518.0             8088   
17       201806               6551.0               457.0             7008   
18       201807               6212.0               595.0             6807   
19       201808               7140.0               843.0             7983   

    Delay_Percentage  Average_Delay  Average_Shipping_time  \
0              17.63          1.994                  2.676   
1              17.16          3.629                  3.206   
2               9.40          5.565                  3.404   
3               6.16          4.181                  3.400   
4               7.92          2.938                  3.346   
5               8.34          2.792                  3.170   
6               7.83          1.782                  3.118   
7               7.58          2.358                  3.127   
8               7.27          2.508                  3.191   
9               7.52          3.064                  3.340   
10              7.93          2.525                  3.576   
11             15.95          3.281                  4.024   
12             10.82          4.188                  3.750   
13             12.65          3.065                  3.886   
14              9.51          3.487                  3.359   
15              7.42          2.567                  3.049   
16              6.40          2.142                  2.701   
17              6.52          2.101                  2.598   
18              8.74          1.959                  2.812   
19             10.56          2.491                  2.909   

    Average_Shipping_estimate  
0                       4.559  
1                       5.196  
2                       7.431  
3                       7.992  
4                       7.768  
5                       7.027  
6                       6.858  
7                       6.831  
8                       7.012  
9                       7.139  
10                      7.015  
11                      7.003  
12                      6.755  
13                      6.878  
14                      6.630  
15                      6.616  
16                      6.129  
17                      6.288  
18                      5.567  
19                      5.200

Now we can plot the graph of delays and average shipping times to use as KPIs for monitor the shipping performance

In [5]:
# Converting the year_months column to datetype to better display on graph axis
order_shipment_delay_per_year_month['year_months'] = pd.to_datetime(order_shipment_delay_per_year_month['year_months'],format='%Y%m').dt.to_period('M')

# Creating a copy of year_months column as strings to better display on the hover tooltips
order_shipment_delay_per_year_month['date'] = order_shipment_delay_per_year_month['year_months'].astype('str')


In [7]:
# Creating the interactive version of the % of orders shipped late every month

# creating the source for plotting in bokeh
graph1_source = ColumnDataSource(order_shipment_delay_per_year_month)

# Defining the figure to draw the plots
p = figure(title="Evolution of Order Dispatch Delay", x_axis_label="Year-Months", y_axis_label="Total Orders Shipped",
                plot_width=990, plot_height=400, tools="", toolbar_location=None, x_axis_type='datetime')


#  Configuring title and axis
p.title.align = 'left'
p.title.text_font_size='12pt'
p.xaxis.ticker.desired_num_ticks = order_shipment_delay_per_year_month['year_months'].nunique()
p.y_range = Range1d(start=0, end=order_shipment_delay_per_year_month['total_shipments'].max())
p.xaxis.major_label_orientation = "vertical"
p.border_fill_color = "whitesmoke"
p.min_border_left = 80


# Creating the line plot of the total shipments of each year-month
a=p.line(x='year_months', y='total_shipments', source = graph1_source,
        color= 'red', line_width=2, legend_label="Total Shipments")


# Creating the second y axis and adding it to the figure
p.extra_y_ranges = {"Percentage" : Range1d(start=0, end=order_shipment_delay_per_year_month['Delay_Percentage'].max()+1)}
p.add_layout(LinearAxis(y_range_name="Percentage", axis_label="% of orders shipped late"), 'right')


# Creating the bar plot with the percentage of orders shipped late
b=p.vbar(x='year_months', top='Delay_Percentage', source=graph1_source, width=dt.timedelta(days=15), alpha=0.5, color= 'blue', legend_label="Delay Percentage", y_range_name="Percentage")


# Adding the hover function to display graphics information
p.add_tools(HoverTool(
    tooltips=[
        ( 'date',   '@date'),
        ( 'orders shipped',  '@total_shipments orders'), # use @{ } for field names with spaces
        ( '% of orders delayed', '@Delay_Percentage %'),
    ],

    # Display a tooltip whenever the cursor is vertically in line with a glyph
    mode='vline',
    renderers=[a]
))

# Configuring y axis color to match the graph
p.yaxis[0].axis_line_color = "red"
p.yaxis[0].major_label_text_color = "red"
p.yaxis[0].major_tick_line_color = "red"
p.yaxis[0].minor_tick_line_color = "red"
p.yaxis[0].axis_label_text_color="red"

p.yaxis[1].axis_line_color = "blue"
p.yaxis[1].major_label_text_color = "blue"
p.yaxis[1].major_tick_line_color = "blue"
p.yaxis[1].minor_tick_line_color = "blue"
p.yaxis[1].axis_label_text_color="blue"

# Adding legend outside the plot
p.add_layout(p.legend[0], 'right')

show(p)


<p style="text-align: center;"> <em>Graph-1: Evolution of Order Dispatch Delay</em> </p>

The graph shows that, even with a continuous increase of shipment number alongside 2017 and stabilizing in 2018, the percentage of orders losing shipment deadlines didn't increase the same way (although it had months with spikes).
Part of the reason for this might be that the shipping process is decentralized and handled by the seller, not the platform.

The increase in December-2017 could be related to the black friday of that year, that concentrates sales on the end of November but most of these sales are processed and shipped in the beginning of December but all these assumptions need more information and a deeper analysis to be confirmed or discarded.

In [8]:
# Let's create a lineplot to visualize the trend of shippment delay and average dispatch times in days

# creating the source for plotting in bokeh
graph2_source = ColumnDataSource(order_shipment_delay_per_year_month)

# Defining the figure to draw the plots
p = figure(title="Behavior of orders shipment, shipping estimates and shipping delays", x_axis_label="Year-Months", y_axis_label="Days",
                plot_width=990, plot_height=400, tools="", toolbar_location=None, x_axis_type='datetime')


#  Configuring title and axis
p.title.align = 'left'
p.title.text_font_size='12pt'
p.xaxis.ticker.desired_num_ticks = 20
p.y_range = Range1d(start=0, end=order_shipment_delay_per_year_month['Average_Shipping_estimate'].max()+1)
p.xaxis.major_label_orientation = "vertical"
p.border_fill_color = "whitesmoke"
p.min_border_left = 80



# Creating the line plot of the average delay for each year-month
a=p.line(x='year_months', y='Average_Delay', source = graph2_source,
        color= 'blue', line_width=2, legend_label="Average Delay")

# Creating the line plot of the Average Shipping Time for each year-month
b=p.line(x='year_months', y='Average_Shipping_time', source = graph2_source,
        color= 'orange', line_width=2, legend_label="Average Shipping Time")

# Creating the line plot of the Average Shipping Estimate for each year-month
c=p.line(x='year_months', y='Average_Shipping_estimate', source = graph2_source,
        color= 'green', line_width=2, legend_label="Average Shipping Estimate")



# Creating 3 circle pĺots just to highlight the selected dots on the Hover ool
t1=p.circle(x='year_months', y='Average_Delay', source = graph2_source, size=7,
                fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")

t2=p.circle(x='year_months', y='Average_Shipping_time', source = graph2_source, size=7,
                fill_color="orange", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")

t3=p.circle(x='year_months', y='Average_Shipping_estimate', source = graph2_source, size=7,
                fill_color="green", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")


# Adding the hover function to display graphics information
p.add_tools(HoverTool(
    tooltips=[
        ( 'date',   '@date'),
        ( 'Average Delay',  '@Average_Delay days'),
        ( 'Average Shipping Time', '@Average_Shipping_time days'),
        ( 'Average Shipping Estimate', '@Average_Shipping_estimate days')
    ],

    # Display a tooltip whenever the cursor is vertically in line with a glyph
    mode='vline',
    renderers=[t1]
))


# Adding legend outside the plot
p.add_layout(p.legend[0], 'right')

show(p)



<p style="text-align: center;"> <em>Graph-2: Behavior of orders shipment, shipping estimates and shipping delays</em> </p>

We can see that the shipping time is usually quite lower than the shipping limit but when an order is delayed, that delay is quite significant (delay is counted as the number of days passed the shipping limit.), reaching over 10 days to be shipped.

Now we need to investigate if these high delays are related to regions or seller specific.

# 2. What are the regions with most delays?


Let's break the shipping behavior by state to investigate possible delay patterns and causes.

In [9]:
# Fetching the order delay data aggregated by state

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/delayed_shipments_by_state_total.sql") as query:
    order_shipment_delay_per_state = pd.read_sql_query(query.read(), params=period, con=connection)

order_shipment_delay_per_state

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


state  Shipped_on_schedule  Shipped_with_delay  total_shipments  \
0     SE                335.0                46.0              381   
1     AL                383.0                52.0              435   
2     PB                528.0                68.0              596   
3     ES               2015.0               226.0             2241   
4     MS                732.0                82.0              814   
5     MA                729.0                81.0              810   
6     PR               5108.0               550.0             5658   
7     RJ              13035.0              1383.0            14418   
8     RN                473.0                50.0              523   
9     TO                284.0                30.0              314   
10    SC               3732.0               387.0             4119   
11    SP              42351.0              4355.0            46706   
12    MG              11758.0              1200.0            12958   
13    CE               1326.0               135.0             1461   
14    BA               3422.0               338.0             3760   
15    PE               1621.0               160.0             1781   
16    DF               2166.0               212.0             2378   
17    RR                 44.0                 4.0               48   
18    RS               5652.0               501.0             6153   
19    PA                982.0                87.0             1069   
20    AC                 85.0                 7.0               92   
21    PI                492.0                40.0              532   
22    RO                253.0                20.0              273   
23    MT                978.0                72.0             1050   
24    GO               2147.0               157.0             2304   
25    AM                156.0                 9.0              165   
26    AP                 77.0                 4.0               81   

    Delay_Percentage  Average_Delay  Average_Shipping_time  \
0              12.07          3.028                  3.608   
1              11.95          3.267                  3.601   
2              11.41          1.880                  3.632   
3              10.08          2.702                  3.431   
4              10.07          2.085                  3.216   
5              10.00          4.097                  3.753   
6               9.72          2.702                  3.304   
7               9.59          3.225                  3.348   
8               9.56          2.946                  3.541   
9               9.55          2.167                  3.476   
10              9.40          3.045                  3.373   
11              9.32          2.913                  3.179   
12              9.26          2.767                  3.241   
13              9.24          3.330                  3.351   
14              8.99          2.789                  3.381   
15              8.98          3.051                  3.273   
16              8.92          3.257                  3.250   
17              8.33          5.514                  3.545   
18              8.14          3.090                  3.264   
19              8.14          3.408                  3.431   
20              7.61          2.110                  3.374   
21              7.52          3.712                  3.210   
22              7.33          1.417                  2.857   
23              6.86          1.954                  3.190   
24              6.81          3.198                  3.086   
25              5.45          1.255                  2.874   
26              4.94          6.237                  3.644   

    Average_Shipping_estimate  
0                       6.803  
1                       6.799  
2                       6.790  
3                       6.683  
4                       6.699  
5                       7.193  
6                       6.581  
7                       6.648  
8                

In [16]:
# Creating the Bokeh function to interact with the plot of shipping delay per state
# this function is necessary for bokeh to run interactive graphs using buttons on jupyter notebooks
# PS: even with this method, bokeh interactive graphics don't show on VSCode notebooks 

def order_shipment_delay_per_state_graph(doc):
     
    df = order_shipment_delay_per_state.sort_values(by='Delay_Percentage', ascending=False).copy()
    
    
    def make_plot(source):
        
        # Defining the figure to draw the plots
        plot = figure(title="Behavior of orders shipment, shipping estimates and shipping delays per state", x_axis_label="State", y_axis_label="Days",
                    plot_width=990, plot_height=400, tools="", toolbar_location=None, x_range=source.data['state'].tolist())
      

        #  Configuring title and axis
        plot.title.align = 'left'
        plot.title.text_font_size='12pt'
        plot.y_range = Range1d(start=0, end=source.data['Delay_Percentage'].max()+1)
        plot.border_fill_color = "whitesmoke"
        plot.min_border_left = 80

        # Creating the second y axis and adding it to the figure
        plot.extra_y_ranges = {"Percentage" : Range1d(start=0, end=source.data['Delay_Percentage'].max()+1)}
        plot.add_layout(LinearAxis(y_range_name="Percentage", axis_label="% of orders shipped late"), 'right')

        # Adjusting the second y-axis color to match the lineplot
        plot.yaxis[1].axis_line_color = "red"
        plot.yaxis[1].major_label_text_color = "red"
        plot.yaxis[1].major_tick_line_color = "red"
        plot.yaxis[1].minor_tick_line_color = "red"
        plot.yaxis[1].axis_label_text_color="red"

        # Creating the line plot of the average delay for each year-month
        plot.line(x='state', y='Average_Delay', source = source,
                color= 'blue', line_width=2, legend_label="Average Delay")

        # Creating the line plot of the Average Shipping Time for each year-month
        plot.line(x='state', y='Average_Shipping_time', source = source,
                color= 'orange', line_width=2, legend_label="Average Shipping Time")

        # Creating the line plot of the Average Shipping Estimate for each year-month
        plot.line(x='state', y='Average_Shipping_estimate', source = source,
                color= 'green', line_width=2, legend_label="Average Shipping Estimate")

        # Plotting the shipping delay percentage
        plot.line(x='state', y='Delay_Percentage', source = source, y_range_name="Percentage",
                color= 'red', line_width=2, legend_label="Delay Percentage")


        # Creating 4 circle pĺots just to highlight the selected dots on the Hover Tool
        t1=plot.circle(x='state', y='Average_Delay', source = source, size=7,
                        fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='state', y='Average_Shipping_time', source = source, size=7,
                        fill_color="orange", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='state', y='Average_Shipping_estimate', source = source, size=7,
                        fill_color="green", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='state', y='Delay_Percentage', source = source, size=7,
                        fill_color="red", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")


        # Adding the hover function to display graphics information
        plot.add_tools(HoverTool(
            tooltips=[
                ( 'state',   '@state'),
                ( 'Delay Percentage', '@Delay_Percentage %'),
                ( 'Average Delay',  '@Average_Delay days'),
                ( 'Average Shipping Time', '@Average_Shipping_time days'),
                ( 'Average Shipping Estimate', '@Average_Shipping_estimate days')
            ],

            mode='vline',
            renderers=[t1]
        ))


        # Adding legend outside the plot
        plot.add_layout(plot.legend[0], 'right')

        return plot


    def update_plot(attr, old, new):
        
                
           sorter = sort_by.value
            
           new_df = order_shipment_delay_per_state.sort_values(by=sorter, ascending=False).copy()
           
           new_source = ColumnDataSource(data=new_df)
        
           plot.x_range.factors=new_source.data['state'].tolist()
            
           source.data = ColumnDataSource.from_df(new_df)
               
        
    
    sort_by = Select(title="Sort By", value="Delay_Percentage", options=['Delay_Percentage', 'Average_Delay', 'Average_Shipping_time', 'Average_Shipping_estimate'])

    source = ColumnDataSource(data=df)

    plot = make_plot(source)

    sort_by.on_change('value', update_plot)
    
    doc.add_root(column(sort_by, plot))

In [17]:
show(order_shipment_delay_per_state_graph, notebook_handle=True)

<p style="text-align: center;"> <em>Graph-3: Behavior of orders shipment, shipping estimates and shipping delays per state</em> </p>

Shipping limit and shipping time are almost flat and parallel lines, wich may indicate that the platform doesn't seem to account order origin to estimate the shipping date (Wich makes sense to guarantee that all orders are processed the same way.). 

But looking into delay percentage and average delay, the highest numbers are concentrated on north and northeast states. This could be related to distance between the seller and buyer or because these states have smaller volume of orders and thus more complex and inefficient freight logistics.

It is interesting to point that the states with highest delay percentages do not have the highest delays in days.

Following, it is presented the evolution of shipping delay statistics for each state:

In [18]:
# Fetching data from all states per year-months

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/delayed_shipments_by_state_monthly.sql") as query:
    order_shipment_delay_per_state_monthly = pd.read_sql_query(query.read(), params=period, con=connection)


# Converting the year_months column to datetype to better display on graph axis
order_shipment_delay_per_state_monthly['year_months'] = pd.to_datetime(order_shipment_delay_per_state_monthly['year_months'],format='%Y%m').dt.to_period('M')

# Creating a copy of year_months column as strings to better display on the hover tooltips
order_shipment_delay_per_state_monthly['date'] = order_shipment_delay_per_state_monthly['year_months'].astype('str')

order_shipment_delay_per_state_monthly

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


year_months state  Shipped_on_schedule  Shipped_with_delay  \
0       2017-01    AC                  4.0                 0.0   
1       2017-01    AL                  0.0                 1.0   
2       2017-01    BA                 20.0                 1.0   
3       2017-01    CE                  6.0                 1.0   
4       2017-01    DF                 12.0                 1.0   
..          ...   ...                  ...                 ...   
528     2018-08    RS                370.0                35.0   
529     2018-08    SC                235.0                22.0   
530     2018-08    SE                 24.0                 2.0   
531     2018-08    SP               3514.0               468.0   
532     2018-08    TO                 17.0                 1.0   

     total_shipments  Delay_Percentage  Average_Delay  Average_Shipping_time  \
0                  4              0.00          0.000                  3.531   
1                  1            100.00          1.371                  5.371   
2                 21              4.76          1.985                  1.753   
3                  7             14.29          0.172                  1.435   
4                 13              7.69          0.019                  2.451   
..               ...               ...            ...                    ...   
528              405              8.64          2.405                  2.959   
529              257              8.56          2.608                  2.820   
530               26              7.69          0.610                  3.396   
531             3982             11.75          2.734                  2.877   
532               18              5.56          1.246                  2.847   

     Average_Shipping_estimate     date  
0                        5.500  2017-01  
1                        4.000  2017-01  
2                        5.333  2017-01  
3                        4.000  2017-01  
4                        4.000  2017-01  
..                         ...      ...  
528                      5.515  2018-08  
529                      5.501  2018-08  
530                      6.338  2018-08  
531                      5.066  2018-08  
532                      6.186  2018-08  

[533 rows x 10 columns]

In [21]:
# Creating the Bokeh function to interact with the plot of shipping delay per state

# "Behavior of orders shipment, shipping estimates and shipping delays for: " state

def order_shipment_delay_per_state_by_year_month(doc):
     
    df = order_shipment_delay_per_state_monthly.copy()
    
    
    def make_plot(source, title):
        
        plot = figure(title=title, x_axis_label="Year-Months", y_axis_label="Days",
                    plot_width=990, plot_height=400, tools="", toolbar_location=None, x_axis_type='datetime')
      

        #  Configuring title and axis
        plot.title.align = 'left'
        plot.title.text_font_size='12pt'
        plot.y_range = Range1d(start=0, end=source.data['Delay_Percentage'].max()+1)
        plot.xaxis.ticker.desired_num_ticks = 20
        plot.xaxis.major_label_orientation = "vertical"
        plot.border_fill_color = "whitesmoke"
        plot.min_border_left = 80

        # Creating the second y axis and adding it to the figure
        plot.extra_y_ranges = {"Percentage" : Range1d(start=0, end=source.data['Delay_Percentage'].max()+1)}
        plot.add_layout(LinearAxis(y_range_name="Percentage", axis_label="% of orders shipped late"), 'right')

        # Adjusting the second y-axis color to match the lineplot
        plot.yaxis[1].axis_line_color = "red"
        plot.yaxis[1].major_label_text_color = "red"
        plot.yaxis[1].major_tick_line_color = "red"
        plot.yaxis[1].minor_tick_line_color = "red"
        plot.yaxis[1].axis_label_text_color="red"

        # Creating the line plot of the average delay for each year-month
        plot.line(x='year_months', y='Average_Delay', source = source,
                color= 'blue', line_width=2, legend_label="Average Delay")

        # Creating the line plot of the Average Shipping Time for each year-month
        plot.line(x='year_months', y='Average_Shipping_time', source = source,
                color= 'orange', line_width=2, legend_label="Average Shipping Time")

        # Creating the line plot of the Average Shipping Estimate for each year-month
        plot.line(x='year_months', y='Average_Shipping_estimate', source = source,
                color= 'green', line_width=2, legend_label="Average Shipping Estimate")

        # Plotting the shipping delay percentage
        plot.line(x='year_months', y='Delay_Percentage', source = source, y_range_name="Percentage",
                color= 'red', line_width=2, legend_label="Delay Percentage")


        # Creating 4 circle pĺots just to highlight the selected dots on the Hover Tool
        t1=plot.circle(x='year_months', y='Average_Delay', source = source, size=7,
                        fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='year_months', y='Average_Shipping_time', source = source, size=7,
                        fill_color="orange", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='year_months', y='Average_Shipping_estimate', source = source, size=7,
                        fill_color="green", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='year_months', y='Delay_Percentage', source = source, size=7,
                        fill_color="red", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white", y_range_name="Percentage")


        # Adding the hover function to display graphics information
        plot.add_tools(HoverTool(
            tooltips=[
                ( 'date',   '@date'),
                ( 'Delay Percentage', '@Delay_Percentage %'),
                ( 'Average Delay',  '@Average_Delay days'),
                ( 'Average Shipping Time', '@Average_Shipping_time days'),
                ( 'Average Shipping Estimate', '@Average_Shipping_estimate days')
            ],

            mode='vline',
            renderers=[t1]
        ))


        # Adding legend outside the plot
        plot.add_layout(plot.legend[0], 'right')

        return plot


    def update_plot(attr, old, new):
        
                
           state = state_select.value
            
           new_df = df[df['state']==state].copy()
           
           plot.title.text = "Behavior of orders shipment, shipping estimates and shipping delays for: " + state
                    
           source.data = ColumnDataSource.from_df(new_df)
        
           plot.extra_y_ranges['Percentage'].end = source.data['Delay_Percentage'].max()+1 
           plot.y_range.end = source.data['Delay_Percentage'].max()+1   
           
               
        
    state="SP"
    
    state_select = Select(title="State", value="SP", options=df['state'].unique().tolist())

    source = ColumnDataSource(data=df[df['state']==state])

    plot = make_plot(source, "Behavior of orders shipment, shipping estimates and shipping delays for: " + state)

    state_select.on_change('value', update_plot)
    
    doc.add_root(column(state_select, plot))

In [22]:
show(order_shipment_delay_per_state_by_year_month, notebook_handle=True)

<p style="text-align: center;"> <em>Graph-4: Behavior of orders shipment, shipping estimates and shipping delays for each state</em> </p>

Next we will explore shipping performance per seller.

# 3. Are there any seller with high delay rate?

A deeper analysis of seller data will be done in another section, but for shipping evaluation, one factor have big impact, only 1% of the sellers are responsible for overt 25% of all items sold in the period. So in order to leverage the impact of those sellers, one way is to categorize them in 3 blocks,
big, medium and small sellers, and analyze each block.

Let's define:

*  Period of data = 2017-01-01 00:00:00 and 2018-08-31 23:59:59
*  Total number of sellers = 2.951 sellers
*  Total number of items sold = 111.120 items sold

| **Seller Type**   | **Classification  Rule in sales** | **Total Sellers** | **Sellers Representation (%)** | **Total Sales** | **Sale Representation (%)** | **sales with shipping delay'** | **delay_percentage** | **Average shipping delay** |
|-------------------|-----------------------------------|-------------------|--------------------------------|-----------------|-----------------------------|--------------------------------|----------------------|----------------------------|
| **Big Seller**    | Over 500 sales                    | 29                | 0.98%                          | 29,168          | 26.25%                      | 2,331                          | 7.99%                | 1.95                       |
| **Medium Seller** | Between 5 and 500 sales           | 1,832             | 62.08%                         | 79,757          | 71.78%                      | 7,643                          | 9.58%                | 3.17                       |
| **Smal Seller**   | less than 5 sales                 | 1,090             | 36.94%                         | 2,195           | 1.97%                       | 271                            | 12.35%               | 5.05                       |
|                   | total                             | 2,951             | 100.00%                        | 111,120         | 100.00%                     | 10,245                         | 9.21%                | 2.94                       |


In [23]:
# Querying seller statistics by seller classification

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/Sellers_classification_by_items_sold.sql") as query:
    seller_category_table = pd.read_sql_query(query.read(), con=connection)

seller_category_table 

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


seller_classification  total_sellers  seller_representation (%)  \
0            Big Seller             29                       0.98   
1         Medium Seller           1861                      63.06   
2          Small Seller           1061                      35.95   

   total_sales  sales_representation (%)  sales with shipping delay  \
0        29168                     26.25                     2331.0   
1        79847                     71.86                     7661.0   
2         2100                      1.89                      248.0   

   delay_percentage  Average shipping delay  
0              7.99                    1.95  
1              9.59                    3.17  
2             11.81                    5.14

In [24]:
# Fetching the order delay data aggregated by seller

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/delayed_shipments_by_seller_category_year_month.sql") as query:
    seller_data_with_category= pd.read_sql_query(query.read(), params=period, con=connection)

# Converting the year_months column to datetype to better display on graph axis
seller_data_with_category['year_months'] = pd.to_datetime(seller_data_with_category['year_months'],format='%Y%m').dt.to_period('M')

# Creating a copy of year_months column as strings to better display on the hover tooltips
seller_data_with_category['date'] = seller_data_with_category['year_months'].astype('str')

seller_data_with_category

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


year_months seller_classification  orders_delayed  total_orders  \
0      2017-01            Big Seller             8.0           131   
1      2017-01         Medium Seller           111.0           565   
2      2017-01          Small Seller             9.0            30   
3      2017-02            Big Seller            50.0           302   
4      2017-02         Medium Seller           226.0          1332   
5      2017-02          Small Seller            18.0            85   
6      2017-03            Big Seller            37.0           686   
7      2017-03         Medium Seller           240.0          2339   
8      2017-03          Small Seller            15.0            80   
9      2017-04            Big Seller            20.0           565   
10     2017-04         Medium Seller           114.0          1776   
11     2017-04          Small Seller            13.0            47   
12     2017-05            Big Seller            39.0           906   
13     2017-05         Medium Seller           283.0          3209   
14     2017-05          Small Seller             9.0            64   
15     2017-06            Big Seller            56.0           916   
16     2017-06         Medium Seller           235.0          2671   
17     2017-06          Small Seller            11.0            56   
18     2017-07            Big Seller           119.0          1240   
19     2017-07         Medium Seller           202.0          2902   
20     2017-07          Small Seller             8.0            58   
21     2017-08            Big Seller           103.0          1354   
22     2017-08         Medium Seller           264.0          3543   
23     2017-08          Small Seller             9.0            61   
24     2017-09            Big Seller            53.0          1356   
25     2017-09         Medium Seller           280.0          3261   
26     2017-09          Small Seller             8.0            74   
27     2017-10            Big Seller            75.0          1492   
28     2017-10         Medium Seller           307.0          3649   
29     2017-10          Small Seller             9.0            72   
30     2017-11            Big Seller           180.0          2380   
31     2017-11         Medium Seller           412.0          5215   
32     2017-11          Small Seller            16.0           114   
33     2017-12            Big Seller           312.0          2212   
34     2017-12         Medium Seller           767.0          4597   
35     2017-12          Small Seller            18.0            76   
36     2018-01            Big Seller           179.0          2454   
37     2018-01         Medium Seller           662.0          5420   
38     2018-01          Small Seller            19.0            86   
39     2018-02            Big Seller           314.0          2058   
40     2018-02         Medium Seller           630.0          5421   
41     2018-02          Small Seller            10.0            70   
42     2018-03            Big Seller           141.0          2306   
43     2018-03         Medium Seller           633.0          5882   
44     2018-03          Small Seller            12.0            78   
45     2018-04            Big Seller           148.0          2156   
46     2018-04         Medium Seller           436.0          5761   
47     2018-04          Small Seller            13.0           126   
48     2018-05            Big Seller           130.0          2195   
49     2018-05         Medium Seller           371.0          5743   
50     2018-05          Small Seller            17.0           150   
51     2018-06            Big Seller           112.0          1579   
52     2018-06         Medium Seller           331.0          5246   
53     2018-06          Small Seller            14.0           183   
54     2018-07            Big Seller           113.0          1419   
55     2018-07         Medium Seller           470.0          5133   
56     2018-0

In [25]:
# Plotting shipping delay information for each seller classification

def shipping_delay_per_seller_category(doc):
     
    df = seller_data_with_category.copy()
    
    
    def make_plot(source, title):
        
        plot = figure(title=title, x_axis_label="Year-Months", y_axis_label="Days",
                    plot_width=990, plot_height=400, tools="", toolbar_location=None, x_axis_type='datetime')
      

        #  Configuring title and axis
        plot.title.align = 'left'
        plot.title.text_font_size='12pt'
        plot.y_range = Range1d(start=0, end=source.data['delay_percentage'].max()+1)
        plot.xaxis.ticker.desired_num_ticks = 20
        plot.xaxis.major_label_orientation = "vertical"
        plot.border_fill_color = "whitesmoke"
        plot.min_border_left = 80

        # Creating the second y axis and adding it to the figure
        plot.extra_y_ranges = {"Percentage" : Range1d(start=0, end=source.data['delay_percentage'].max()+1)}
        plot.add_layout(LinearAxis(y_range_name="Percentage", axis_label="% of orders shipped late"), 'right')

        # Adjusting the second y-axis color to match the lineplot
        plot.yaxis[1].axis_line_color = "red"
        plot.yaxis[1].major_label_text_color = "red"
        plot.yaxis[1].major_tick_line_color = "red"
        plot.yaxis[1].minor_tick_line_color = "red"
        plot.yaxis[1].axis_label_text_color="red"

        # Creating the line plot of the average delay for each year-month
        plot.line(x='year_months', y='average_delay', source = source,
                color= 'blue', line_width=2, legend_label="Average Delay")

        # Plotting the shipping delay percentage
        plot.line(x='year_months', y='delay_percentage', source = source, y_range_name="Percentage",
                color= 'red', line_width=2, legend_label="Delay Percentage")


        # Creating 4 circle pĺots just to highlight the selected dots on the Hover Tool
        t1=plot.circle(x='year_months', y='average_delay', source = source, size=7,
                        fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        
        plot.circle(x='year_months', y='delay_percentage', source = source, size=7,
                        fill_color="red", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white", y_range_name="Percentage")


        # Adding the hover function to display graphics information
        plot.add_tools(HoverTool(
            tooltips=[
                ( 'date',   '@date'),
                ( 'Delay Percentage', '@delay_percentage %'),
                ( 'Average Delay',  '@average_delay days')  
            ],

            mode='vline',
            renderers=[t1]
        ))


        # Adding legend outside the plot
        plot.add_layout(plot.legend[0], 'right')

        return plot


    def update_plot(attr, old, new):
        
                
           seller_category = seller_category_select.value
            
           new_df = df[df['seller_classification']==seller_category].copy()
           
           plot.title.text = "Evolution of Shipping Delay Percentage and Time for " + seller_category
                    
           source.data = ColumnDataSource.from_df(new_df)
        
           plot.extra_y_ranges['Percentage'].end = source.data['delay_percentage'].max()+1 
           plot.y_range.end = source.data['delay_percentage'].max()+1   
           
               
        
    seller_category="Big Seller"
    
    seller_category_select = Select(title="Seller Category", value="Big Seller", options=df['seller_classification'].unique().tolist())

    source = ColumnDataSource(data=df[df['seller_classification']==seller_category])

    plot = make_plot(source, "Evolution of Shipping Delay Percentage and Time for " + seller_category)

    seller_category_select.on_change('value', update_plot)
    
    doc.add_root(column(seller_category_select, plot))

In [26]:
show(shipping_delay_per_seller_category, notebook_handle=True)

<p style="text-align: center;"> <em>Graph-5: Evolution of Shipping Delay Percentage and Time for each seller category</em> </p>

It is observable that all seller classifications had an increase in shipments delayed on the period between 2017-12 and 2018-02.
Although small sellers demonstrated a reduction tendency in their order delays, they still take the highest time to ship orders with delay.
Also, this average delay on small sellers oscillated a lot alongside the observed period.

Comparing Big and Medium sellers, both have similar average delay but medium sellers showed less volatile behavior since their numbers variated less 
during the observed period.

To finalize this analysis, let's extract delivery information and summarize the performance of the platform and their seller partners.

# 4. How many orders were delivered with delay?

Since there is no information about the carriers responsible for each delivery, we cannot assess their individual performance.
This way, we will limit the analysis to point out important statistics such as:
* Average delivery time;
* Average delivery delay (days past delivery limit);
* Delivery delay by state (state of order origin);
* The impact of seller and buyer being of different states in those statistics;

#### Average delivery time and Average delivery delay

Let's present delivery delay evolution during the observed period.

Since we are going to explore delivery statistics, we need to consider as reference date the "order_delivered_customer_date" data.
If we use "order_delivered_carrier_date" or even "order_purchase_timestamp" we will encounter orders that don't have a delivery date information
(orders with status of shipped or cancelled).

This information is important and needs to be assessed with stakeholders before beginning any analysis because the numbers may diverge from other analysis
that need to use a different date as base(like income analysis that use date of payment as reference or marketing analysis using purchase timestamp to
analyze sales on a specific marketing campaign period and etc) and sometimes those divergences may cause noise on meetings and even invalidate the report results until a proper explanation is given.

We will use the following constraints for the delivery analysis:
* Use "order_delivered_customer_date" as date reference for data aggregation
* consider only orders with "order_delivered_customer_date" filled (not null or blank)
* consider only orders delivered to customers between '2017-01-01 00:00:00' and '2018-08-31 23:59:59'
* consider only orders  with "order_status" = delivered

In [27]:
# Fetching the order delivery delay data aggregated by year-month

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/delivery_delay_statistics_by_year_month.sql") as query:
    order_delivery_data= pd.read_sql_query(query.read(), params=period, con=connection)
    
# Converting the year_months column to datetype to better display on graph axis
order_delivery_data['year_months'] = pd.to_datetime(order_delivery_data['year_months'],format='%Y%m').dt.to_period('M')

# Creating a copy of year_months column as strings to better display on the hover tooltips
order_delivery_data['date'] = order_delivery_data['year_months'].astype('str')

order_delivery_data

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


year_months  delivered_on_schedule  delivered_with_delay  total_deliveries  \
0      2017-01                  326.0                   0.0               326   
1      2017-02                 1562.0                   3.0              1565   
2      2017-03                 2654.0                  70.0              2724   
3      2017-04                 1946.0                 126.0              2072   
4      2017-05                 3961.0                 240.0              4201   
5      2017-06                 3511.0                 129.0              3640   
6      2017-07                 3782.0                 151.0              3933   
7      2017-08                 4772.0                 128.0              4900   
8      2017-09                 4309.0                 187.0              4496   
9      2017-10                 4965.0                 250.0              5215   
10     2017-11                 5142.0                 270.0              5412   
11     2017-12                 7454.0                 865.0              8319   
12     2018-01                 6684.0                 735.0              7419   
13     2018-02                 6204.0                 419.0              6623   
14     2018-03                 6706.0                1241.0              7947   
15     2018-04                 7402.0                1597.0              8999   
16     2018-05                 7665.0                 501.0              8166   
17     2018-06                 7239.0                 648.0              7887   
18     2018-07                 6474.0                 107.0              6581   
19     2018-08                 8406.0                 979.0              9385   

    delay_percentage  Average_Delay  average_order_lifetime  \
0               0.00            NaN                   8.545   
1               0.19          1.169                  10.452   
2               2.57          7.300                  11.874   
3               6.08          9.323                  13.032   
4               5.71          7.579                  12.892   
5               3.54          9.465                  11.657   
6               3.84          7.801                  11.260   
7               2.61          9.116                  10.795   
8               4.16         27.060                  12.308   
9               4.79          7.384                  11.674   
10              4.99          8.512                  11.858   
11             10.40          6.282                  14.126   
12              9.91         12.648                  15.221   
13              6.33          8.815                  14.594   
14             15.62          7.132                  15.321   
15             17.75         11.540                  15.126   
16              6.14         15.738                  11.643   
17              8.22          8.036                  11.591   
18              1.63         13.694                   9.051   
19             10.43          3.983                   8.699   

    average_delivery_duration     date  
0                       5.886  2017-01  
1                       7.482  2017-02  
2                       8.546  2017-03  
3                       9.573  2017-04  
4                       9.456  2017-05  
5                       8.588  2017-06  
6                       8.108  2017-07  
7                       7.658  2017-08  
8                       9.151  2017-09  
9                       8.359  2017-10  
10                      8.309  2017-11  
11                     10.321  2017-12  
12                     11.389  2018-01  
13                     10.838  2018-02  
14                     11.678  2018-03  
15                     11.902  2018-04  
16                      8.958  2018-05  
17                      8.880  2018-06  
18                      6.385  2018-07  
19                      5.781  2018-08

In [29]:
# Creation of a lineplot to visualize the trend of delivery delay and average delivery time in days

# creating the source for plotting in bokeh
graph6_source = ColumnDataSource(order_delivery_data)

# Defining the figure to draw the plots
p = figure(title="Behavior of orders delivery, delivery estimates and delivery delays", x_axis_label="Year-Months", y_axis_label="Days",
                plot_width=990, plot_height=400, tools="", toolbar_location=None, x_axis_type='datetime')


#  Configuring title and axis
p.title.align = 'left'
p.title.text_font_size='12pt'
p.xaxis.ticker.desired_num_ticks = 20
p.y_range = Range1d(start=0, end=30)
p.xaxis.major_label_orientation = "vertical"
p.border_fill_color = "whitesmoke"
p.min_border_left = 80


# Creating the second y axis and adding it to the figure
p.extra_y_ranges = {"Percentage" : Range1d(start=0, end=30)}
p.add_layout(LinearAxis(y_range_name="Percentage", axis_label="% of orders Delivered late"), 'right')

# Adjusting the second y-axis color to match the lineplot
p.yaxis[1].axis_line_color = "red"
p.yaxis[1].major_label_text_color = "red"
p.yaxis[1].major_tick_line_color = "red"
p.yaxis[1].minor_tick_line_color = "red"
p.yaxis[1].axis_label_text_color="red"


# Creating the line plot of the average delivery delay for each year-month
a=p.line(x='year_months', y='Average_Delay', source = graph6_source,
        color= 'blue', line_width=2, legend_label="Average Delay")

# Creating the line plot of the Average Delivery Time for each year-month
b=p.line(x='year_months', y='average_delivery_duration', source = graph6_source,
        color= 'orange', line_width=2, legend_label="Average Delivery Time")


# Creating the line plot of the Average Order Lifetime for each year-month
c=p.line(x='year_months', y='delay_percentage', source = graph6_source, y_range_name="Percentage",
        color= 'red', line_width=2, legend_label="Delivery Delay Percentage")




# Creating 3 circle pĺots just to highlight the selected dots on the Hover ool
t1=p.circle(x='year_months', y='Average_Delay', source = graph6_source, size=7,
                fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")

t2=p.circle(x='year_months', y='average_delivery_duration', source = graph6_source, size=7,
                fill_color="orange", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")


t3=p.circle(x='year_months', y='delay_percentage', source = graph6_source, size=7,
                fill_color="red", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")



# Adding the hover function to display graphics information
p.add_tools(HoverTool(
    tooltips=[
        ( 'date',   '@date'),
        ( 'Delay Percentage', '@delay_percentage %'),
        ( 'Average Delay',  '@Average_Delay days'),
        ( 'Average Shipping Time', '@average_delivery_duration days')
    ],

    # Display a tooltip whenever the cursor is vertically in line with a glyph
    mode='vline',
    renderers=[t1]
))


# Adding legend outside the plot
p.add_layout(p.legend[0], 'right')


show(p)

<p style="text-align: center;"> <em>Graph-6: Behavior of orders delivery, delivery estimates and delivery delays</em> </p>

Breaking the graphic in 2 parts, first we have a delivery duration with a slightly tendency of increase followed by a reduction tendency, fluctuating most of the time between 5 and 10 days. 
Second, the delay percentage also shows a tendency of increase in 2017 but followed by a strong fluctuation in 2018.

Finally, there is an abnormally high delay in 2017-09 that need to be inspected and validated, wich we will do first before get back to analyze the delay percentage fluctuation.

In [30]:
# extracting the data from 2017-09 to investigate the high delay

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/order_delivered_in_2017-09.sql") as query:
    order_delivered_september_2017= pd.read_sql_query(query.read(), con=connection)
order_delivered_september_2017

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


order_id order_status order_purchase_timestamp  \
0    ca07593549f1816d26a572e06dc1eab6    delivered      2017-02-21 23:31:27   
1    2fe324febf907e3ea3f2aa9650869fa5    delivered      2017-03-13 20:17:10   
2    285ab9426d6982034523a855f55a885e    delivered      2017-03-08 22:47:40   
3    440d0d17af552815d15a9e41abe49359    delivered      2017-03-07 23:59:51   
4    c27815f7e3dd0b926b58552628481575    delivered      2017-03-15 23:23:17   
..                                ...          ...                      ...   
170  3cba655faae78feacf27f20900c776e0    delivered      2017-09-06 20:10:17   
171  43f1558f9eec3596990432207c65c146    delivered      2017-08-28 20:03:11   
172  91575acf2a356af85894e1c3b1612e81    delivered      2017-09-06 13:07:48   
173  9413addabc2268ea995d4e968a185ab3    delivered      2017-08-14 19:45:18   
174  950863e2e170cafe0f1363b6debbf73c    delivered      2017-09-07 20:33:34   

       order_approved_at order_delivered_carrier_date  \
0    2017-02-23 02:35:15          2017-03-08 13:47:46   
1    2017-03-13 20:17:10          2017-03-17 07:23:53   
2    2017-03-08 22:47:40          2017-03-09 08:53:20   
3    2017-03-09 01:11:33          2017-03-15 13:00:08   
4    2017-03-15 23:23:17          2017-03-21 14:50:25   
..                   ...                          ...   
170  2017-09-06 20:25:11          2017-09-12 19:12:55   
171  2017-08-29 04:35:50          2017-08-30 19:42:06   
172  2017-09-06 13:24:07          2017-09-12 13:21:23   
173  2017-08-16 03:30:17          2017-08-18 15:34:43   
174  2017-09-07 20:45:13          2017-09-11 21:37:46   

    order_delivered_customer_date order_estimated_delivery_date     delay  
0             2017-09-19 14:36:39           2017-03-22 00:00:00  181.6083  
1             2017-09-19 17:00:07           2017-04-05 00:00:00  167.7083  
2             2017-09-19 14:00:04           2017-04-06 00:00:00  166.5833  
3             2017-09-19 15:12:50           2017-04-07 00:00:00  165.6333  
4             2017-09-19 17:14:25           2017-04-10 00:00:00  162.7181  
..                            ...                           ...       ...  
170           2017-09-26 17:05:08           2017-09-26 00:00:00    0.7118  
171           2017-09-08 16:25:57           2017-09-08 00:00:00    0.6840  
172           2017-09-29 13:46:06           2017-09-29 00:00:00    0.5736  
173           2017-09-05 12:55:24           2017-09-05 00:00:00    0.5382  
174           2017-09-20 10:41:36           2017-09-20 00:00:00    0.4451  

[175 rows x 8 columns]

Exploring the data from September we can see that there are 48 entries with over 15 days of delay on delivery and even 21 orders with over 100 days of delay.

Analyzing those 21 entries and validating the calculations manually, the calculations proceed and all 21 orders have all information fields filled, indicating they are regular orders with no apparent error in the columns (like purchase timestamp older than delivery date and etc)
Unfortunately, there are no extra data to be used for us to understand better the reasons behind this behavior.

In [31]:
# Fetching the order delivery data aggregated by state

period = ('2017-01-01 00:00:00', '2018-08-31 23:59:59')

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/delivery_delay_statistics_per_state.sql") as query:
    order_delivery_data_by_state= pd.read_sql_query(query.read(), params=period, con=connection)

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [35]:
# Creating the plot of delivery delay per state

def order_delivery_delay_per_state_graph(doc):
     
    df = order_delivery_data_by_state.sort_values(by='delivery_delay_percentage', ascending=False).copy()
    
    
    def make_plot(source):
        
        plot = figure(title="Behavior of orders delivery, delivery estimates and delivery delays per state", x_axis_label="State", y_axis_label="Days",
                    plot_width=990, plot_height=400, tools="", toolbar_location=None, x_range=source.data['state'].tolist())
      

        #  Configuring title and axis
        plot.title.align = 'left'
        plot.title.text_font_size='12pt'
        plot.y_range = Range1d(start=0, end=source.data['average_delay'].max()+1)
        plot.border_fill_color = "whitesmoke"
        plot.min_border_left = 80

        # Creating the second y axis and adding it to the figure
        plot.extra_y_ranges = {"Percentage" : Range1d(start=0, end=source.data['average_delay'].max()+1)}
        plot.add_layout(LinearAxis(y_range_name="Percentage", axis_label="% of orders delivered late"), 'right')

        # Adjusting the second y-axis color to match the lineplot
        plot.yaxis[1].axis_line_color = "red"
        plot.yaxis[1].major_label_text_color = "red"
        plot.yaxis[1].major_tick_line_color = "red"
        plot.yaxis[1].minor_tick_line_color = "red"
        plot.yaxis[1].axis_label_text_color="red"

        # Creating the line plot of the average delivery delay for each state
        plot.line(x='state', y='average_delay', source = source,
                color= 'blue', line_width=2, legend_label="Average Delay")

        # Creating the line plot of the Average Delivery Time for each state
        plot.line(x='state', y='average_delivery_time', source = source,
                color= 'orange', line_width=2, legend_label="Average Delivery Time")

        # Creating the line plot of the Average Delivery Estimate for each state
        plot.line(x='state', y='average_delivery_estimate', source = source,
                color= 'green', line_width=2, legend_label="Average Delivery Estimate")

        # Plotting the delivery delay percentage
        plot.line(x='state', y='delivery_delay_percentage', source = source, y_range_name="Percentage",
                color= 'red', line_width=2, legend_label="Delay Percentage")


        # Creating 4 circle pĺots just to highlight the selected dots on the Hover Tool
        t1=plot.circle(x='state', y='average_delay', source = source, size=7,
                        fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='state', y='average_delivery_time', source = source, size=7,
                        fill_color="orange", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='state', y='average_delivery_estimate', source = source, size=7,
                        fill_color="green", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")

        plot.circle(x='state', y='delivery_delay_percentage', source = source, size=7,
                        fill_color="red", hover_fill_color="firebrick", hover_alpha=0.7,
                        line_color=None, hover_line_color="white")


        # Adding the hover function to display graphics information
        plot.add_tools(HoverTool(
            tooltips=[
                ( 'state',   '@state'),
                ( 'Delay Percentage', '@delivery_delay_percentage %'),
                ( 'Average Delay',  '@average_delay days'),
                ( 'Average Delivery Time', '@average_delivery_time days'),
                ( 'Average Delivery Estimate', '@average_delivery_estimate days')
            ],

            mode='vline',
            renderers=[t1]
        ))


        # Adding legend outside the plot
        plot.add_layout(plot.legend[0], 'right')

        return plot


    def update_plot(attr, old, new):
        
                
           sorter = sort_by.value
            
           new_df = order_delivery_data_by_state.sort_values(by=sorter, ascending=False).copy()
           
           new_source = ColumnDataSource(data=new_df)
        
           plot.x_range.factors=new_source.data['state'].tolist()
            
           source.data = ColumnDataSource.from_df(new_df)
               
        
    
    sort_by = Select(title="Sort By", value="delivery_delay_percentage", options=['delivery_delay_percentage', 'average_delay', 'average_delivery_time', 'average_delivery_estimate'])

    source = ColumnDataSource(data=df)

    plot = make_plot(source)

    sort_by.on_change('value', update_plot)
    
    doc.add_root(column(sort_by, plot))

In [36]:
show(order_delivery_delay_per_state_graph, notebook_handle=True)

<p style="text-align: center;"> <em>Graph-7: Behavior of orders delivery, delivery estimates and delivery delays per state</em> </p>

Delivery statistics shows the highest delivery times are concentrated mostly on the northern region (wich makes sense as north is the most remote region of Brazil and most sellers are concentrated in the southeast region) followed by the northeast region.

Looking into delay information, northeast states have higher percentage of orders delivered after the deadline. Interesting point on delay is that most north region have the lowest percentage of delivery delay even having the worst logistics conditions (bad roads, huge distances, most logistics are made by plane or boat). This way, the common conclusion that "delivering orders to distant or difficult access places tend to delay more often" diverge from the data we have.

For the average delay (time past deadline for delayed deliveries), AP, RR, AM and AC have the highest numbers (they are also the most remote states of the country). This could be related to distance or difficult access but we don't have enough carrier and delivery data to test this hypotesis.

# 5. How shipping delay impacts delivery performance?

Let's see if missing shipment deadlines also increases delivery delays. For that, let's compare the percentage of delivery delay of orders shipped with and without delay.

In [38]:
# Fetching the order delivery data aggregated by year_month and with shipping delay flag

with open("2-Queries/1-Shipment_and_delivery/Ship_and_delivery_status.sql") as query:
    delivery_percentage_with_shipment_status= pd.read_sql_query(query.read(), params=period, con=connection)

# Converting the year_months column to datetype to better display on graph axis
delivery_percentage_with_shipment_status['year_months'] = pd.to_datetime(delivery_percentage_with_shipment_status['year_months'],format='%Y%m').dt.to_period('M')

# Creating a copy of year_months column as strings to better display on the hover tooltips
delivery_percentage_with_shipment_status['date'] = delivery_percentage_with_shipment_status['year_months'].astype('str')
    
delivery_percentage_with_shipment_status

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


year_months  Ship_ok_Delivery_ok  Ship_Late_Delivery_ok  \
0      2017-01              82.2086                17.7914   
1      2017-02              83.1949                16.6134   
2      2017-03              87.2614                10.1689   
3      2017-04              88.7548                 5.1641   
4      2017-05              88.5980                 5.6891   
5      2017-06              89.9451                 6.5110   
6      2017-07              89.3466                 6.8141   
7      2017-08              90.3673                 7.0204   
8      2017-09              89.9021                 5.9386   
9      2017-10              88.8207                 6.3854   
10     2017-11              89.0429                 5.9682   
11     2017-12              80.2140                 9.3881   
12     2018-01              81.9248                 8.1682   
13     2018-02              83.6328                10.0408   
14     2018-03              77.0605                 7.3235   
15     2018-04              77.0641                 5.1895   
16     2018-05              89.4808                 4.3840   
17     2018-06              86.8011                 4.9829   
18     2018-07              90.2143                 8.1599   
19     2018-08              83.7826                 5.7858   

    Ship_ok_Delivery_Late  Ship_Late_Delivery_Late  tot_delivered     date  
0                  0.0000                   0.0000            326  2017-01  
1                  0.1278                   0.0639           1565  2017-02  
2                  1.2115                   1.3583           2724  2017-03  
3                  3.8610                   2.2201           2072  2017-04  
4                  3.9990                   1.7139           4201  2017-05  
5                  2.5549                   0.9890           3640  2017-06  
6                  2.5172                   1.3221           3933  2017-07  
7                  1.8980                   0.7143           4900  2017-08  
8                  3.0694                   1.0899           4496  2017-09  
9                  3.2598                   1.5340           5215  2017-10  
10                 3.3075                   1.6814           5412  2017-11  
11                 6.7917                   3.6062           8319  2017-12  
12                 6.6451                   3.2619           7419  2018-01  
13                 4.1975                   2.1289           6623  2018-02  
14                11.8787                   3.7373           7947  2018-03  
15                14.8016                   2.9448           8999  2018-04  
16                 4.8249                   1.3103           8166  2018-05  
17                 6.2128                   2.0033           7887  2018-06  
18                 1.0789                   0.5470           6581  2018-07  
19                 6.1161                   4.3154           9385  2018-08

In [41]:
# Creating the plot to compare the percentage of orders shipped and delivered late

# creating the dataset to be plotted
graph8_source = ColumnDataSource(delivery_percentage_with_shipment_status)

# Defining the figure to draw the plots
p = figure(title='Orders distributed by delay on delivery and shipment', x_axis_label="Year-Months", y_axis_label="% of Total Orders",
                plot_width=990, plot_height=400, tools="", toolbar_location=None, x_axis_type='datetime')


#  Configuring title and axis
p.title.align = 'left'
p.title.text_font_size='12pt'
p.xaxis.ticker.desired_num_ticks = 20
p.y_range = Range1d(start=0, end=20)
p.xaxis.major_label_orientation = "vertical"
p.border_fill_color = "whitesmoke"
p.min_border_left = 80



# Creating the line plot of the orders shipped on schedule but delivered late
a=p.line(x='year_months', y='Ship_ok_Delivery_Late', source = graph8_source,
        color= 'blue', line_width=2, legend_label="Shipped OK Delivered late")

# Creating the line plot of the orders shipped late but delivered on schedule
b=p.line(x='year_months', y='Ship_Late_Delivery_ok', source = graph8_source,
        color= 'orange', line_width=2, legend_label="Shipped Late Delivered OK")


# Creating the line plot of the orders shipped late and delivered late
c=p.line(x='year_months', y='Ship_Late_Delivery_Late', source = graph8_source,
        color= 'red', line_width=2, legend_label="Shipped Late Delivered Late")




# Creating 3 circle pĺots just to highlight the selected dots on the Hover ool
t1=p.circle(x='year_months', y='Ship_ok_Delivery_Late', source = graph8_source, size=7,
                fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")

t2=p.circle(x='year_months', y='Ship_Late_Delivery_ok', source = graph8_source, size=7,
                fill_color="orange", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")


t3=p.circle(x='year_months', y='Ship_Late_Delivery_Late', source = graph8_source, size=7,
                fill_color="red", hover_fill_color="firebrick", hover_alpha=0.7,
                line_color=None, hover_line_color="white")



# Adding the hover function to display graphics information
p.add_tools(HoverTool(
    tooltips=[
        ( 'date',   '@date'),
        ( 'Shipped OK Delivered late', '@Ship_ok_Delivery_Late %'),
        ( 'Shipped Late Delivered OK',  '@Ship_Late_Delivery_ok %'),
        ( 'Shipped Late Delivered Late', '@Ship_Late_Delivery_Late %')
    ],

    # Display a tooltip whenever the cursor is vertically in line with a glyph
    mode='vline',
    renderers=[t1]
))


# Adding legend outside the plot
p.add_layout(p.legend[0], 'right')


show(p)

<p style="text-align: center;"> <em>Graph-8: Orders distributed by delay on delivery and shipment</em> </p>

# 6. Is there any relation between delivery delay and the distance between seller and customer?

To answer this question, a few key data are needed:
*   Buyer and seller location (present on  this dataset)
*   Timestamp for shipping and delivery (also present)
*   descriptive data for each delivery such as transport type, presence of incidents, processing times in border crossing etc. (none of this data is available)

Since we don't have any information on delivery process and its possible problems, making an unbiased analysis becomes complex.

Over 90% of sellers are concentrated on southeast and south regions, this way, states from those regions should have the lowest numbers of delay but in fact, the delivery delay is fairly linear through states from southeast, south and northeast. This alone doesn't disprove the hypothesis.

Delay percentage also should be higher on remote regions but the north region possess the lowest percentage of orders delivered after deadlines. Wich also conflicts with the hypothesis but is not enough to disprove it.

So, having conflicting information from delivery delay percentage and delay duration indicates that is not possible to have an accurate conclusion without looking into extra data.

# 7. How does freight value behaves among states?

Let's finalize this analysis by looking into freight distribution. The most direct approach would be calculating the average freight for each origin-destination pair. We will create an interactive panel to show this information filtered per state.

Let's show the average freight value a customer would pay depending on the state he/she lives:

In [42]:
# Fetching the freight information aggregated by buyer state

period = ('2017-01-01 00:00:00', '2018-08-31 23:59:59')

# Loading the query
with open("2-Queries/1-Shipment_and_delivery/freight_value_per_buyer_state.sql") as query:
    freight_value_per_buyer_state= pd.read_sql_query(query.read(), params=period, con=connection)
freight_value_per_buyer_state

/home/hbeltrao/Hugo/py_environments/data_env/lib/python3.10/site-packages/pandas/io/sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


customer_state  avg_freight  tot_orders
0              RR    44.101458          48
1              PB    42.853708         596
2              RO    41.330549         273
3              AC    40.073370          92
4              PI    39.044586         532
5              MA    38.371272         810
6              TO    37.307962         314
7              SE    36.823753         381
8              AL    35.940253         435
9              PA    35.883761        1069
10             RN    35.703193         523
11             AP    34.160494          81
12             AM    33.205394         165
13             CE    32.705222        1461
14             PE    32.698871        1781
15             MT    27.970362        1050
16             BA    26.404875        3760
17             MS    23.359017         814
18             GO    22.527617        2304
19             ES    22.022030        2241
20             RS    21.641984        6153
21             SC    21.491282        4119
22             DF    21.075008        2378
23             RJ    20.941118       14418
24             MG    20.625723       12958
25             PR    20.491748        5658
26             SP    15.121402       46706

In [44]:
# Plotting freight value distribution by state
graph9_source = ColumnDataSource(data=freight_value_per_buyer_state)

plot = figure(title="Average Freight Value per state", x_axis_label="State", y_axis_label="R$",
    plot_width=990, plot_height=400, tools="", toolbar_location=None, x_range=graph9_source.data['customer_state'].tolist())
      

#  Configuring title and axis
plot.title.align = 'left'
plot.title.text_font_size='12pt'
plot.y_range = Range1d(start=0, end=graph9_source.data['avg_freight'].max()+1)
plot.border_fill_color = "whitesmoke"
plot.min_border_left = 80


# Creating the line plot of the average delay for each year-month
plot.line(x='customer_state', y='avg_freight', source = graph9_source,
color= 'blue', line_width=2, legend_label="Average Freight Value")


# Creating circle pĺots just to highlight the selected dots on the Hover Tool
t1=plot.circle(x='customer_state', y='avg_freight', source = graph9_source, size=7,
fill_color="blue", hover_fill_color="firebrick", hover_alpha=0.7,
line_color=None, hover_line_color="white")


# Adding the hover function to display graphics information
plot.add_tools(HoverTool(
    tooltips=[
( 'State',   '@customer_state'),
( 'Average Freight', 'R$ @avg_freight'),
( 'Total Orders',  '@tot_orders orders')
    ],

mode='vline',
renderers=[t1]
))
    
show(plot)

<p style="text-align: center;"> <em>Graph-9: Orders distributed by delay on delivery and shipment</em> </p>

Looking to the graph, there is a significant difference between freight value from southeast and south states compared to the north and northeast states.
Since most sellers and buyers are from south and southeast, this result make sense.

# 7. Conclusions

Summarising  the key information  extracted  in the  analysis  we  have:  
  
*  The increase  in  sales  on the  platform  didn't cause an  increase  in  delay  of orders  shipments;  
*  Although  shipment  time  estimate have  decreased  alongside the  period, the real  shipment  time  didn't  followed  this  evolution;  
*  Orders of northeast  customers  tend to  miss  shipping  deadlines more often than  orders  of  customers  from other  regions:  

	* In general, orders  take, on average, the same amount of time to be shipped independently of the customer region;

	* To  avoid  bias in conclusions regarding shipping  prioritization, further analysis is needed such as verifying other prioritization  variables like order  value, quantity of order items, freight value  etc; 
	
* Small  sellers  tend to  lose  shipment  dedline  more often and the  time  after the  deadline  also is longer for  small  sellers;  
* Shipping  late doesn't reflect in  an  increase  on delivery delay;  
* The conflict  on existing  data  and  lack  of  extra  data  about delivery delay and  distance  between  seller  and  buyer unfeasibling unbiased analysis and  conclusion;
* Freight  is cheaper for south and southeast buyers;


To turn all this information in a useful tool, it could be created a recommendation system that when a client search for a product, it will return a list of sellers with the best product price alongside with freight and delivery time estimates. Usually, this information is displayed inside the product web page and not on the search list.